torch/_dynamo/variables/torch.py

```diff
@@ -406,36 +407,20 @@ def call_function(
                 )
             else:
                 unimplemented(f"torch.from_numpy(<{type(t)}>)")
-        elif len(args) > 0 and isinstance(args[0], TensorWithTFOverrideVariable):
-            # This code block implements inlining the __torch_function__
-            # override of a tensor.
-
-            tensor_with_tf_override = args[0]
-
-            # TODO(future PR): make this implement the full __torch_function__ API
-            # instead of assuming the relevant override is in the first argument.
-            args[0] = args[0].tensor_variable
-
-            unwrapped = TensorWithTFOverrideVariable.inline_torch_function_unwrapped(
-                tx,
-                self,
-                tensor_with_tf_override.orig_tensor_variable_source,
-                tensor_with_tf_override.subclass_torch_function__func,
-                tensor_with_tf_override.subclass_type,
-                options,
-                args,
-                kwargs,
-            )
+        elif can_dispatch_torch_function(tx, args, kwargs):
+            # continue the assumption that args[0] is the torch function
+            # tensor until we fully trace the base torch function impl
 
+            unwrapped = dispatch_torch_function(tx, self, args, kwargs)
             # The wrapping here follows the logic in
             # `torch.Tensor.__torch_function__`.
             if self.value in torch.overrides.get_default_nowrap_functions():
                 return unwrapped
-            return TensorWithTFOverrideVariable(
+            return TensorWithTFOverrideVariable.create(
+                tx,
                 unwrapped,
-                tensor_with_tf_override.orig_tensor_variable_source,
-                tensor_with_tf_override.subclass_torch_function__func,
-                tensor_with_tf_override.subclass_type,
+                args[0].torch_function_fn,
+                args[0].subclass_type,
```

Comment:
I literally reread this line and had the exact same question I asked in https://github.com/pytorch/pytorch/pull/109556/files/97ae25444562832e77e0cc78b9b369e593f39732#diff-7bd43c6b174845f3450be42c8460ed021d1b39a98510cbea5d4e7e3a0b9d8d4f

PLEASE improve the comment. If you are looking for a concrete suggestion, here is what I would have written

```
# TODO: It's not correct to always dispatch on args[0]; with multiple subclasses the dispatch
# may be on the second or later argument. Fix this to respect what dispatch_torch_function says 
# the dispatch should be.
```

Except that last sentence is not quite right because dispatch_torch_function doesn't seem to tell you which args you should be dispatching on.

Reply:
Actually, reading `dispatch_torch_function` carefully, I guess you are planning to use `unwrapped.torch_function_fn` instead? Can you just literally say this here? It would make code review so much easier.

If we're adding TODOs, there are two more things that stick out like a sore thumb here: (1) the default nowrap functions logic shouldn't be here, it should be in the default torch function implementation (a user torch function can override the behavior and wrap things that aren't default wrapped), and (2) you shouldn't be unconditionally wrapping in TensorWithTFOverrideVariable because a torch function can return a plain tensor, removing the subclass entirely.

Reply:
I added TODOs for each of these cases